google/cloud/forseti/scanner/audit/retention_rules_engine.py

```diff
@@ -0,0 +1,318 @@
+
+    def find_buckets_violations(self, buckets_lifecycle, force_rebuild=False):
+        """Determine whether bucket lifecycle violates rules.
+
+        Args:
+            buckets_lifecycle (retention_bucket.RetentionBucket): Object
+                containing lifecycle data
+            force_rebuild (bool): If True, rebuilds the rule book. This will
+                reload the rules definition file and add the rules to the book.
+
+        Returns:
+             generator: A generator of rule violations.
+        """
+        if self.rule_book is None or force_rebuild:
+            self.build_rule_book()
+
+        violations = itertools.chain()
+
+        resource_rules = self.rule_book.get_resource_rules(_APPLY_TO_BUCKETS)
+
+        recent_bucket = resource_util.create_resource(
+            resource_id=buckets_lifecycle.name,
+            resource_type=_APPLY_TO_BUCKETS)
+
+        resource_ancestors = (relationship.find_ancestors(
+            recent_bucket, buckets_lifecycle.full_name))
+        for related_resources in resource_ancestors:
+            rules = resource_rules[related_resources]
+            for rule in rules:
+                violations = itertools.chain(
+                    violations,
+                    rule.find_buckets_violations(buckets_lifecycle))
+
+        return set(violations)
+
+
+class RetentionRuleBook(bre.BaseRuleBook):
+    """The RuleBook for Retention resources."""
+
+    def __init__(self, rule_defs=None):
+        """Initialization.
+
+        Args:
+            rule_defs (dict): rule definitons
+        """
+        super(RetentionRuleBook, self).__init__()
+        self._rules_sema = threading.BoundedSemaphore(value=1)
+
+        self.resource_rules_map = {
+            applies_to: collections.defaultdict(set)
+            for applies_to in _APPLY_TO_RESOURCES}
+        if not rule_defs:
+            self.rule_defs = {}
+        else:
+            self.rule_defs = rule_defs
+            self.add_rules(rule_defs)
+
+    def add_rules(self, rule_defs):
+        """Add rules to the rule book
+
+        Args:
+            rule_defs (dict): rule definitions dictionary
+        """
+        for (i, rule) in enumerate(rule_defs.get('rules', [])):
+            self.add_rule(rule, i)
+
+    def add_rule(self, rule_def, rule_index):
+        """Add a rule to the rule book.
+
+        Args:
+            rule_def (dict): A dictionary containing rule definition
+                properties.
+            rule_index (int): The index of the rule from the rule definitions.
+                Assigned automatically when the rule book is built.
+        """
+        self._rules_sema.acquire()
+        try:
+            applies_to = rule_def.get('applies_to', None)
+            if applies_to is None:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Lack of applies_to in rule {}'.format(rule_index))
+            if type(applies_to).__name__ != 'list':
```

Comment:
We generally haven't been checking types of rule fields, so you can remove the check. I think type checks would best be done with a schema file: #1962 .

Also, the check can be written as: `isinstance(applies_to, list)`